feat(color): allow all UI controls for widgets when full screen

philmoz opened the pull request:
Extends the Lua LVGL API to allow buttons, text and number edit, sliders etc to be used on widgets when the widget is in full screen mode.

Example widget script:
```Lua

local opts = {
}

local wgt
local exitApp = false

local function create(zone, opts)
  wgt = { zone=zone, opts=opts }
  return wgt
end

local function update(wgt, opts)
  wgt.opts = opts
  exitApp = false

  if lvgl == nil then return end

  lvgl.clear()

  local lyt

  if (lvgl.isFullScreen()) then
    lyt = {
      {type="button", text="Close", x=100, y=100,
          press=(
              function()
                  lvgl.confirm({title="Exit", message="Really exit?",
                      confirm=(function() exitApp = true end)
                  })
              end)},
    }
  else
    lyt = {
      {type="label", text="Enter full screen for UI", x=100, y=100},
    }
  end

  lvgl.build(lyt)
end

local function background(wgt)
end

local function refresh(wgt, event, touchState)
  if (exitApp) then
    lcd.exitFullScreen()
  end

  if lvgl == nil then
    lcd.drawText(0, 0, "Lvgl support required", COLOR_THEME_WARNING)
  end
end

return { name="Lvgl App", options=opts, create=create, update=update, refresh=refresh, background=background, useLvgl=true }
```

wimalopaan commented:
Many thanks for your great work! 

Most stuff was already discussed on discord.

Just not to get lost I mention the following issues I actually see:

* `slider`: function `get` not called (only once the object gets visible)
* maybe new UI class `momentaryButton` as discussed: https://discord.com/channels/839849772864503828/842693696629899274/1331892836303638582

philmoz commented:
Slider and toggle switch fixed.
You can use any number in place of the PAD_xx constants so not sure a PAD_XL is really needed.

A note on the lvgl.getContext() function.
For widgets it returns the same table that was created with the first 'create()' function call. This allows functions that get or set properties in UI elements to access the internal script data without having to pass the table to every function call.
For stand alone scripts it returns nil.
This was requested by @offer-shmuely.

wimalopaan commented:
> Slider and toggle switch fixed. 

Confirm with my test app. Thanks!!!

> You can use any number in place of the PAD_xx constants so not sure a PAD_XL is really needed.

Ok, maybe need only a docu fix then.

> 
> A note on the lvgl.getContext() function. For widgets it returns the same table that was created with the first 'create()' function call. This allows functions that get or set properties in UI elements to access the internal script data without having to pass the table to every function call. For stand alone scripts it returns nil. This was requested by @offer-shmuely.

Cool, will also have a look at that.

wimalopaan commented:
> Added 'momentaryButton' control:

Wow! Absolute great work. Many thanks.

First tests are successful ;-)

wimalopaan commented:
Just an idea for a small improvement (in my eyes):

the UI classes `button` and `momentaryButton` share the same scrren layout - they are visually indistinguishable. I would prefer to have different looks for them, to say: the `momentaryButton` should get another layout (maybe italic font or not a rounded rectangle).

That said, I would it find useful to override the default colors (from theme). The `color` property for e.g. `button` is ignored. Maybe there could be `textColor`, `fgColor`, `bgColor` that override the theme defaults.

offer-shmuely commented:
Great addition For ease of documentation, it will be better as buttonMomentarySo it will be sorted alphabetically (docs/web site/auto completion)Can the edgeTx used that button instead of the regular in the numbers editor (-10, -1, +1, +10)?—————————Offer ShmuelyOn 24 Jan 2025, at 16:58, Wilhelm ***@***.***> wrote:﻿
Just an idea for a small improvement (in my eyes):
the UI classes button and momentaryButton share the same screen layout - they are visually indistinguishable. I would prefer to have different looks for them, to say: the momentaryButton should get another layout (maybe italic font or not a rounded rectangle).
That said, I would it find useful to override the default colors (from theme). The color property for e.g. button is ignored. Maybe there could be textColor, fgColor, bgColor that override the theme defaults.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

philmoz commented:
> Great addition For ease of documentation, it will be better as buttonMomentarySo it will be sorted alphabetically

@wimalopaan - the momentary button was your idea, do you have any preference for how it is named? Now would be the time to change it.

philmoz commented:
> the UI classes `button` and `momentaryButton` share the same screen layout - they are visually indistinguishable. I would prefer to have different looks for them, to say: the `momentaryButton` should get another layout (maybe italic font or not a rounded rectangle).

@JimB40 - do you have any thoughts on this?

wimalopaan commented:
> > Great addition For ease of documentation, it will be better as buttonMomentarySo it will be sorted alphabetically
> 
> @wimalopaan - the momentary button was your idea, do you have any preference for how it is named? Now would be the time to change it.

I would still vote for `momentaryButton`.

wimalopaan commented:
> > the UI classes `button` and `momentaryButton` share the same screen layout - they are visually indistinguishable. I would prefer to have different looks for them, to say: the `momentaryButton` should get another layout (maybe italic font or not a rounded rectangle).
> 
> @JimB40 - do you have any thoughts on this?

The `momentaryButton` is new and up to now not used in EdgeTx, so there would be no interference with existing UI code. And I think a (small) visual distinction would be useful.

wimalopaan commented:
> That said, I would it find useful to override the default colors (from theme). The `color` property for e.g. `button` is ignored. Maybe there could be `textColor`, `fgColor`, `bgColor` that override the theme defaults.

Does this make sense?

wimalopaan commented:
I did some intensive tests with this PR and it looks really well!
The _only_ caveat is that in widget-mode of TBS Agent Lite (0.99 for LUA 5.3) the roller does not work anymore (I used https://github.com/EdgeTX/edgetx/commit/8c192f242e709ec7768adc78ab523c9da9c271ab for testing).

wimalopaan commented:
Since `lcd.exitFullScreen()` does not work for 'lvgl' widgets: how can widgets leave fullscreen mode by themselves?

philmoz commented:
> I did some intensive tests with this PR and it looks really well! The _only_ caveat is that in widget-mode of TBS Agent Lite (0.99 for LUA 5.3, @JimB40 ) the roller does not work anymore (I used [8c192f2](https://github.com/EdgeTX/edgetx/commit/8c192f242e709ec7768adc78ab523c9da9c271ab) for testing).

Not sure what you mean - widgets only get rotary encode events when in full screen mode.

philmoz commented:
> Since `lcd.exitFullScreen()` does not work for `lvgl` widgets: how can widgets leave fullscreen mode by themselves?

lcd.exitFullScreen() is still available.

wimalopaan commented:
> > I did some intensive tests with this PR and it looks really well! The _only_ caveat is that in widget-mode of TBS Agent Lite (0.99 for LUA 5.3, @JimB40 ) the roller does not work anymore (I used [8c192f2](https://github.com/EdgeTX/edgetx/commit/8c192f242e709ec7768adc78ab523c9da9c271ab) for testing).
> 
> Not sure what you mean - widgets only get rotary encoder events when in full screen mode.

Sorry for being unclear: the TBS Agent lite can be used as stand-alone script (works well) and as widget. If installed as widget and switched to full-screen, the roller does not work. Well, unfortunately it is only available as compiled lua code.

wimalopaan commented:
> > Since `lcd.exitFullScreen()` does not work for `lvgl` widgets: how can widgets leave fullscreen mode by themselves?
> 
> lcd.exitFullScreen() is still available. It's only the drawing functions that do not work.

Mmh, I tried to use it in the `back` function, but that did not work (maybe I'm doing something wrong).

philmoz commented:
> > > Since `lcd.exitFullScreen()` does not work for `lvgl` widgets: how can widgets leave fullscreen mode by themselves?
> > 
> > 
> > lcd.exitFullScreen() is still available. It's only the drawing functions that do not work.
> 
> Mmh, I tried to use it in the `back` function, but that did not work (maybe I'm doing something wrong).

It will work if called in the run() function; but not elsewhere when using lvgl.
Probably need to add a separate lvgl.exitFullScreen() function.

philmoz commented:
Rotary encoder events should work again for full screen widgets that do not use LVGL.

I've added lvgl.exitFullScreen() function.
Note: I also renamed lvgl.isFullscreen to lvgl.isFullScreen for consistency.

philmoz commented:
Added three new properties to the 'button' and 'momentaryButton' controls:
- 'color' - sets the background color for the button
- 'textColor' - sets the text color for the button
- 'cornerRadius' - sets the roundness of the corners (0 = square)

If you set 'w' equal to 'h' and 'cornerRadius' to w/2 then you will get a round button.

If don't set these they default to the firmware style values.

wimalopaan commented:
> Rotary encoder events should work again for full screen widgets that do not use LVGL.
> 
> I've added lvgl.exitFullScreen() function. Note: I also renamed lvgl.isFullscreen to lvgl.isFullScreen for consistency.

Looks good to me now ;-)

wimalopaan commented:
> Added three new properties to the 'button' and 'momentaryButton' controls:
> 
> * 'color' - sets the background color for the button
> * 'textColor' - sets the text color for the button
> * 'cornerRadius' - sets the roundness of the corners (0 = square)
> 
> If you set 'w' equal to 'h' and 'cornerRadius' to w/2 then you will get a round button.
> 
> If don't set these they default to the firmware style values.

Very cool! 
Tested that successfully.
Many thanks.

offer-shmuely commented:
thanks for all the improvements!!

offer-shmuely commented:
Do do you think that the +/- button can be momentary-buttons?
So a long press will continue to change the value?

![image](https://github.com/user-attachments/assets/94e9c7f4-126c-4191-b207-c5c6fb00eac3)

philmoz commented:
The buttons already had auto-repeat enabled.

wimalopaan commented:
The `active` function of UI class `slider` does not work.

wimalopaan commented:
The `font` property does not work for UI class `button` or `momentaryButton`

wimalopaan commented:
The UI class `slider` renders a horizontal slider. But there is no vertical slider. It would be useful to have also a UI class `vSlider` or `verticalSlider` or add a property `orientation` that can be set to `vertical` or `horizontal`.

philmoz commented:
Fixed the 'active' state for the slider.
Fixed the 'font' property for buttons.
Added 'verticalSlider' control - note you need to set the 'h' property to control size.

wimalopaan commented:
> Fixed the 'active' state for the slider. Fixed the 'font' property for buttons. Added 'verticalSlider' control - note you need to set the 'h' property to control size.

Your pace is breathtaking!
Tested positiv on simulator (tests on radio will follow).
Many thanks again ;-)

wimalopaan commented:
It would be cool to extend the function of the sliders in two ways:

* add an optional detent at some value via property `detent`, maybe need a property `mid` for the position at which the detent occurs
* add an option to make the slider self-neutralizing to a position, see above: property: `mid`.

wimalopaan commented:
The UI class `slider` does not reacz to the `color` property. Maybe it would be possible to give the slider a _color_ that overrides the theme color. This color could be the color of the knob of the slider ...

wimalopaan commented:
Just an idea: would it be possible to make a _round_ slider, like  a _gauge_ but as active UI-element? Well, this would be new to EdgeTx and probably to much effort.

philmoz commented:
> It would be cool to extend the function of the sliders in two ways:
> 
> * add an optional detent at some value via property `detent`, maybe need a property `mid` for the position at which the detent occurs
> * add an option to make the slider self-neutralizing to a position, see above: property: `mid`.

Not sure I understand these. The slider is not a physical control so how could it have a detent. You can also reset the slider position at any time in your Lua code.

philmoz commented:
> Just an idea: would it be possible to make a _round_ slider, like a _gauge_ but as active UI-element? Well, this would be new to EdgeTx and probably to much effort.

You can create one using arcs, lines and triangles.

philmoz commented:
> The UI classes `switch` and `source` display _all_ switches or sources. Sometimes it may be useful to present to the user only a filtered list of switches (or sources), e.g. only logical switches. Therefore it would be nice to have a `filter` property (function) that is called with a switch (or source) index and return `true` / `false` to decide if it is listed to the user.

How would you know what each switch or source index represents - the values can vary by radio type and are subject to change over time.

Internally the Source filtering is done by type (stick, pot, trim, etc) where the code uses a bit mask with the allowed types. This could be exposed to the Lua API.

We don't have an equivalent for switches at this point.

wimalopaan commented:
> > It would be cool to extend the function of the sliders in two ways:
> > 
> > * add an optional detent at some value via property `detent`, maybe need a property `mid` for the position at which the detent occurs
> > * add an option to make the slider self-neutralizing to a position, see above: property: `mid`.
> 
> Not sure I understand these. The slider is not a physical control so how could it have a detent. 

I'm sure you now the behaviour that some PC window-manager have: some windows snap into position directly aligned one to the other. I you want to move one of them, you habe to move the mouse some more pixel than normal. Don't know hoe to describe that better here, but I'm sure, you know what I mean. Sometimes this is called a "physics simulation".

You can also reset the slider position at any time in your Lua code.

What I meant with self-neutralizing: I you move the slider and release the event, the slider moves back to the neutral position: like a stick that i smoved and then let it go by itself to the neutral position.

wimalopaan commented:
> > Just an idea: would it be possible to make a _round_ slider, like a _gauge_ but as active UI-element? Well, this would be new to EdgeTx and probably to much effort.
> 
> You can create one using arcs, lines and triangles.

Sure. Was just an idea.

wimalopaan commented:
> > The UI classes `switch` and `source` display _all_ switches or sources. Sometimes it may be useful to present to the user only a filtered list of switches (or sources), e.g. only logical switches. Therefore it would be nice to have a `filter` property (function) that is called with a switch (or source) index and return `true` / `false` to decide if it is listed to the user.
> 
> How would you know what each switch or source index represents - the values can vary by radio type and are subject to change over time.

You are right: I errorneous assumed that the switch index would be also valid to e.g. `model.getLogicalSwitch()` but that isn't true.  
Mmh: I don't know if there is another way to filter the switches ...

philmoz commented:
I've added a 'filter' property to the source picker control. By default it shows everything.

You can change it by either combining the constants below to pick source types, or subtracting them from SRC_ALL if you just want to remove a few.
```Lua
-- Inputs, switches and trims only
{type="source", filter=lvgl.SRC_INPUT|lvgl.SRC_SWITCH|lvgl.SRC_TRIM}

-- Remove 'Invert' option and global variables
{type="source", filter=lvgl.SRC_ALL-lvgl.SRC_CLEAR-lvgl.SRC_INVERT-lvgl.SRC_GVAR}
```

Source types:
- lvgl.SRC_ALL - everything
- lvgl.SRC_INPUT - inputs
- lvgl.SRC_STICK - analog sticks
- lvgl.SRC_POT - pots and sliders
- lvgl.SRC_SWITCH - switches
- lvgl.SRC_CHANNEL - outputs
- lvgl.SRC_TRIM - trims
- lvgl.SRC_LOGICAL_SWITCH - logical switches
- lvgl.SRC_GVAR - global variables
- lvgl.SRC_LUA - custom mix script outputs
- lvgl.SRC_OTHER - MIN, MAX, tx battery, timers, etc
- lvgl.SRC_HELI - heli values
- lvgl.SRC_TRAINER - trainer channels
- lvgl.SRC_TELEM - telemetry sensors
- lvgl.SRC_CLEAR - special value to control the 'Clear' button option
- lvgl.SRC_INVERT - special value to control the 'Invert' button option

philmoz commented:
I've added the 'filter' to the switch source as well.

Switch types:

lvgl.SW_ALL - everything
lvgl.SW_SWITCH- switches
lvgl.SW_TRIM - trims
lvgl.SW_LOGICAL_SWITCH - logical switches
lvgl.SW_TELEM - telemetry sensors
lvgl.SW_OTHER - ON, ONE, etc
lvgl.SW_CLEAR - special value to control the 'Clear' button option

wimalopaan commented:
> I've added the 'filter' to the switch source as well.

Very cool! Thank you.
Looks good to me.

philmoz commented:
The 'color' property on the slider control will now set the color of the inside part of the knob, it also sets the tick marks to the same color.

philmoz commented:
> > > It would be cool to extend the function of the sliders in two ways:
> > > 
> > > * add an optional detent at some value via property `detent`, maybe need a property `mid` for the position at which the detent occurs
> > > * add an option to make the slider self-neutralizing to a position, see above: property: `mid`.
> > 
> > 
> > Not sure I understand these. The slider is not a physical control so how could it have a detent.
> 
> I'm sure you now the behaviour that some PC window-manager have: some windows snap into position directly aligned one to the other. I you want to move one of them, you habe to move the mouse some more pixel than normal. Don't know hoe to describe that better here, but I'm sure, you know what I mean. Sometimes this is called a "physics simulation".

Sorry, I don't understand how this would apply to a slider control.
 
> > You can also reset the slider position at any time in your Lua code.
> 
> What I meant with self-neutralizing: I you move the slider and release the event, the slider moves back to the neutral position: like a stick that i smoved and then let it go by itself to the neutral position.

Unlikely, the event handling for the slider is buried within the lvgl code.

wimalopaan commented:
> The 'color' property on the slider control will now set the color of the inside part of the knob, it also sets the tick marks to the same color.

Perfect, works as expected.

wimalopaan commented:
> Sorry, I don't understand how this would apply to a slider control.

Well, you have to move the event-location (finger) a certain amount before the slider follows to the new postion. Like the distance from the detent-position is a measure of the force that drags the slider (like a spring). 

But: no problem, was just an idea.

wimalopaan commented:
> Unlikely, the event handling for the slider is buried within the lvgl code.

Ok.
Would it be possible to implement a double-tap event with a callback? Or simply move the slider to the neutral position if a double-tap happens?

wimalopaan commented:
Actually I get some problems when using two(!) instances of the same widget. After creating some UI pages the widgets stopped working correctly (e.g. button does not go into checked state) and the radio can't be switched off of model is still connected because the ENTER touch button does not work anymore.

I'll have to dig deeper into the problem ...

wimalopaan commented:
> I'll have to dig deeper into the problem ...

Using a DEBUG Build the problem has gone. But I see `LUA instructionsPercent 123%` which might be the problem ...

gagarinlg commented:
> > I'll have to dig deeper into the problem ...
> 
> Using a DEBUG Build the problem has gone. But I see `LUA instructionsPercent 123%` which might be the problem ...

You can enable a debug function in lv_conf.h that shows which areas are redrawn. Maybe that helps debugging the high load

wimalopaan commented:
Ok, thanks für the info.

With a `DEBUG` I'm not able to reproduce the problem, but I see some LUA instruction percent up to 145%. That doen not look good. Maybe I have to split the page building into several passes and add the UI element-groups one after another ...

philmoz commented:
> Ok, thanks für the info.
> 
> With a `DEBUG` I'm not able to reproduce the problem, but I see some LUA instruction percent up to 145%. That doen not look good. Maybe I have to split the page building into several passes and add the UI element-groups one after another ...

Can you share the script code please.

wimalopaan commented:
> > Ok, thanks für the info.
> > With a `DEBUG` I'm not able to reproduce the problem, but I see some LUA instruction percent up to 145%. That doen not look good. Maybe I have to split the page building into several passes and add the UI element-groups one after another ...
> 
> Can you share the script code please.

The link to the widget: https://github.com/wimalopaan/LUA/tree/main/WIDGETS/lvglMultiSw
The directory name must be `/WIDGETS/lvglMultiSw` as in the repo.

philmoz commented:
> > > Ok, thanks für the info.
> > > With a `DEBUG` I'm not able to reproduce the problem, but I see some LUA instruction percent up to 145%. That doen not look good. Maybe I have to split the page building into several passes and add the UI element-groups one after another ...
> > 
> > 
> > Can you share the script code please.
> 
> The link to the widget: https://github.com/wimalopaan/LUA/tree/main/WIDGETS/lvglMultiSw The directory name must be `/WIDGETS/lvglMultiSw` as in the repo.
> 
> Install 2 (or more) instances and alter the `Address` option.
> 
> The make them fullscreen, go to settings page and edit e.g. a name. After that try to tap and check a button on the controls page.

Tested in simulator and on TX16S and T15 radios and it works fine for me.
Noticed a couple of things:
- on initial setup both widgets are saving to the MODEL_xx_0.lua file so they overwrite each other. After reboot they switch to the correct file based on the address setting.
- the button checked state is saved; but is not restored on re-entering full screen so all the buttons look unchecked even though internally you consider them checked. It takes two taps to get them showing checked again.

wimalopaan commented:
> Tested in simulator and on TX16S and T15 radios and it works fine for me.

Did you test in a DEBUG build firmware?

> * on initial setup both widgets are saving to the MODEL_xx_0.lua file so they overwrite each other. After reboot they switch to the correct file based on the address setting.

Yes, and already fixed.

> * the button checked state is saved; but is not restored on re-entering full screen so all the buttons look unchecked even though internally you consider them checked. It takes two taps to get them showing checked again.

Thanks for reporting and testing.

I had the problem on non-DEBUG TX16s and non-DEBUG X12s firmware, with DEBUG that never happens.

wimalopaan commented:
In a debug build the LUA instruction limit is disabled, so the error can not happen.

But in release build an error message should be displayed.

philmoz commented:
Both radios were release builds.
Retested with the firmware built by this PR and both are fine.

I selected a simple model with no other scripts. Set the view layout to 2x1 and added the widget into both spots, set addresses to 20 and 40.

wimalopaan commented:
On my X12s the problem occurs in Release builds with two instances of this script and two other scripts. 

I think, I shall decompose the page construction into several steps to reduce the LUA instruction counter.

philmoz commented:
> On my X12s the problem occurs in Release builds with two instances of this script and two other scripts.
> 
> I think, I shall decompose the page construction into several steps to reduce the LUA instruction counter.

I think this is a bug in the way the Lua instruction limit is being handled. Looking at it now.

philmoz commented:
I've added a change for the instruction limit issue - can you see if it works better for you.

wimalopaan commented:
Thanks again! 
Now, I don't see any `instructionsPercent` messages in `DEBUG` builds anymore, and no more _hangs_ in release builds. 

(The only thing that reaches too high `instructionsPercent` is TBS Agent Lite (as widget), which goes > 230%)

wimalopaan commented:
> > Unlikely, the event handling for the slider is buried within the lvgl code.
> 
> Ok. Would it be possible to implement a double-tap event with a callback? Or simply move the slider to the neutral position if a double-tap happens?

Do you see a chance to implement that?

pfeerick commented:
Is this PR pretty much baked now? Anything major still outstanding?

philmoz commented:
I think this is good to merge. If there are any more enhancements or bugs we can create a new PR.

wimalopaan commented:
Absolutely!

pfeerick commented:
LGTM on TX16... will be great to see some new scripts using this. Thank you @wimalopaan for all your testing, and for the widget scripts you have worked on already making use of this :)

wimalopaan commented:
I have to come back to this wonderful PR ;-) 

Actually I'm doing some more graphics intensive widgets: here, I like to draw some graphical primitives like rectangles and triangles and arc with rapidly changing parameters. For the `arc` one could use values or functions to determine the parameters, for `rectangle` and `triangle` only values are possible.
I would find function very handy in my situation, because it avoids using a `name` for each graphical primitive. And it looks like it would be not too complicated to add this ... maybe symmetry in the API may also be an argument for that.

philmoz commented:
> Actually I'm doing some more graphics intensive widgets: here, I like to draw some graphical primitives like rectangles and triangles and arcs with rapidly changing parameters. For the `arc` one could use values or functions to determine the parameters, for `rectangle` and `triangle` only values are possible.

There is an open PR for Lua changes - #5926 
Can you add a list of which properties you are wanting to change to that PR so they don't get lost please.
